haystack/components/rankers/hugging_face_tei.py

```diff
@@ -0,0 +1,247 @@
+        :param token: The Hugging Face token to use as HTTP bearer authorization.
+            Check your HF token in your [account settings](https://huggingface.co/settings/tokens).
+        :param max_retries: Maximum number of retry attempts for failed requests (default: 3).
+        :param retry_status_codes: List of HTTP status codes that will trigger a retry.
+            When None, HTTP 408, 418, 429 and 503 will be retried (default: None).
+        """
+        # Construct the full rerank endpoint
+        self.url = url
+        self.top_k = top_k
+        self.timeout = timeout
+        self.token = token
+        self.max_retries = max_retries
+        self.retry_status_codes = retry_status_codes
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Serializes the component configuration to a dictionary for saving or logging.
+
+        :returns: A dict containing the component's initialization parameters.
+        """
+        return default_to_dict(
+            self,
+            url=self.url,
+            top_k=self.top_k,
+            timeout=self.timeout,
+            token=self.token.to_dict() if self.token else None,
+            max_retries=self.max_retries,
+            retry_status_codes=self.retry_status_codes,
+        )
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "HuggingFaceTEIRanker":
+        """
+        Deserializes the component from a configuration dictionary.
+
+        :param data: Configuration dict produced by `to_dict()`.
+
+        :returns: An initialized `HuggingFaceTEIRanker` instance.
+        """
+        deserialize_secrets_inplace(data["init_parameters"], keys=["token"])
+        return default_from_dict(cls, data)
+
+    def _compose_response(
+        self, result: Union[Dict[str, str], List[Dict[str, Any]]], top_k: Optional[int], documents: List[Document]
+    ) -> Dict[str, List[Document]]:
+        """
+        Processes the API response into a structured format.
+
+        :param result: The raw response from the API.
+
+        :returns: A dict with key `documents` mapping to the reranked top documents.
+        """
+        if isinstance(result, dict) and "error" in result:
+            error_type = result.get("error_type", "UnknownError")
+            error_msg = result.get("error", "No additional information.")
+            raise RuntimeError(f"HuggingFaceTEIRanker API call failed ({error_type}): {error_msg}")
+
+        # Ensure we have a list of score dicts
+        if not isinstance(result, list):
+            raise RuntimeError("Unexpected response format from HuggingFaceTEIRanker API.")
+
+        # Determine number of docs to return
+        final_k = min(top_k or self.top_k, len(result))
+
+        # Select and return the top_k documents
+        ranked_docs = []
+        for item in result[:final_k]:
+            index: int = item["index"]
+            documents[index].score = item["score"]
```

Comment:
i would copy the document to avoid modifying it in-place (`copy.copy` is sufficient)